crates/provider-blobstore-nats/src/provider.rs

```diff
@@ -0,0 +1,391 @@
+        .context("failed to serve provider exports")
+    }
+
+    /// Build a [`NatsBlobstoreProvider`] from [`HostData`]
+    pub fn from_host_data(host_data: &HostData) -> NatsBlobstoreProvider {
+        let config = NatsConnectionConfig::from_link_config(&host_data.config, &host_data.secrets);
+        if let Ok(config) = config {
+            NatsBlobstoreProvider {
+                default_config: config,
+                ..Default::default()
+            }
+        } else {
+            warn!("failed to build NATS connection configuration, falling back to default");
+            NatsBlobstoreProvider::default()
+        }
+    }
+
+    /// Attempt to connect to NATS url (with JWT credentials, if provided)
+    async fn connect(
+        &self,
+        cfg: NatsConnectionConfig,
+    ) -> anyhow::Result<async_nats::jetstream::context::Context> {
+        let mut opts = match (cfg.auth_jwt, cfg.auth_seed) {
+            (Some(jwt), Some(seed)) => {
+                let seed = KeyPair::from_seed(&seed).context("failed to parse seed key pair")?;
+                let seed = Arc::new(seed);
+                async_nats::ConnectOptions::with_jwt(jwt, move |nonce| {
+                    let seed = seed.clone();
+                    async move { seed.sign(&nonce).map_err(async_nats::AuthError::new) }
+                })
+            }
+            (None, None) => async_nats::ConnectOptions::default(),
+            _ => bail!("must provide both jwt and seed for jwt authentication"),
+        };
+        if let Some(tls_ca) = &cfg.tls_ca {
+            opts = add_tls_ca(tls_ca, opts)?;
+        } else if let Some(tls_ca_file) = &cfg.tls_ca_file {
+            let ca = fs::read_to_string(tls_ca_file)
+                .await
+                .context("failed to read TLS CA file")?;
+            opts = add_tls_ca(&ca, opts)?;
+        }
+
+        // Get the cluster_uri with proper default
+        let uri = cfg.cluster_uri.unwrap_or(DEFAULT_NATS_URI.to_string());
+
+        // Connect to the NATS server
+        let client = opts
+            .name("NATS Object Store Provider")
+            .connect(uri.clone())
+            .await?;
+
+        // Get the JetStream context based on js_domain
+        let jetstream = if let Some(domain) = &cfg.js_domain {
+            async_nats::jetstream::with_domain(client.clone(), domain.clone())
+        } else {
+            async_nats::jetstream::new(client.clone())
+        };
+
+        debug!("opened NATS JetStream: {:?}", jetstream);
+        debug!("NATS Connection Configuration: {:?}", client);
+
+        // Return the handle to the opened NATS Object store
+        Ok(jetstream)
+    }
+
+    /// Helper function to lookup and return the NATS JetStream connection handle, and container storage
+    /// configuration, using the client component's context.
+    /// This ensures consistent implementation across all functions that need to get the NATS Blobstore.
+    pub(crate) async fn get_blobstore(
```

Comment:
```suggestion
    /// This ensures consistent implementation across all functions that need to get the NATS Blobstore.
    #[instrument(level = "debug", skip_all)]
    pub(crate) async fn get_blobstore(
```

Small suggestion to ensure this function is traced, let me know if you left it out purposefully though